[6.2 🍒][Dependency Scanning] Emit a detailed warning diagnostic on Clang module variant discovery

artemcm opened the pull request:
Cherry-pick of https://github.com/swiftlang/swift/pull/81313
-----------------------
- **Explanation**: In expectation, this should never happen. Such a situation means that within the same scanning action, Clang Dependency Scanner has produced two different variants of the same module. We are currently hunting down the rare cases where it does, seemingly due to differences in Clang Scanner direct by-name queries and transitive header lookup queries and potential other causes. The code-path that this change modifies is only possible to trigger in a case that would already guarantee a build failure down-the-line.

- **Scope**: Rare cases where dependency scanning ends up in a situation where it "discovers" two variants of the same Clang module

- **Risk**: Low, this change affects only a code-path that would lead to a downstream build failure.

- **Issue**: rdar://149230376

- **Reviewed By**: @beccadax 

- **Original PR**: https://github.com/swiftlang/swift/pull/81313

artemcm committed:
[Dependency Scanning] Emit a detailed error diagnostic on Clang module variant discovery

In expectation, this should never happen. Such a situation means that within the same scanning action, Clang Dependency Scanner has produced two different variants of the same module. This is not supposed to happen, but we are currently hunting down the rare cases where it does, seemingly due to differences in Clang Scanner direct by-name queries and transitive header lookup queries.

artemcm commented:
@swift-ci test

nkcsgexi reviewed (approved):
We've seen multiple cases where builds complain about PCM being out of date. The root cause of these failures is mostly project configurations which take a lot of engineering time to track down. Having more detailed and explicit error messages like this helps us track down those issues more efficiently, thus I consider this to be a high value change to be cherry-picked. Additionally, this pure additive diagnostic logics are quite safe too.

nkcsgexi commented on include/swift/AST/DiagnosticsFrontend.def:
```diff
@@ -618,5 +618,15 @@ ERROR(ast_format_requires_dump_ast,none,
 ERROR(unknown_dump_ast_format,none,
       "unknown format '%0' requested with '-dump-ast-format'", (StringRef))
 
+ERROR(dependency_scan_unexpected_variant, none,
```
Can we downgrade this from error to a warning? I suspect this would happen semi-frequently that a fatal error is probably an overkill.

nkcsgexi replied:
We should discuss more about the implication with @cyndyishida and other folks. Please hold off merging before that.

artemcm committed:
Downgrade the scanning module variant diagnostic to a warning, for now.

There is concern that this diagnostic may be breaking in some cases on its own.

nkcsgexi commented:
Thank you for downgrading to warning Artem! It's then safe enough to be merged for 6.2.

artemcm commented:
@swift-ci test macOS platform
